apm-sniffer/apm-sdk-plugin/pulsar-plugin/src/main/java/org/apache/skywalking/apm/plugin/pulsar/PulsarProducerInterceptor.java

```diff
@@ -0,0 +1,110 @@
+ *  3. Inject the context to {@link Message#getProperties()}
+ *  4. Create {@link SendCallbackEnhanceRequiredInfo} with <code>ContextManager.capture()</code> and set the
+ *     callback enhanced instance skywalking dynamic field to the created required info.
+ *  5. Stop the exit span when <code>sendAsync</code> method finished.
+ * </pre>
+ *
+ * @author penghui
+ */
+public class PulsarProducerInterceptor implements InstanceMethodsAroundInterceptor {
+
+    public static final String OPERATE_NAME_PREFIX = "Pulsar/";
+    public static final String PRODUCER_OPERATE_NAME_SUFFIX = "/Producer";
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                             MethodInterceptResult result) throws Throwable {
+        if (allArguments.length > 0 && allArguments[0] != null) {
+            ProducerEnhanceRequiredInfo requiredInfo = (ProducerEnhanceRequiredInfo) objInst.getSkyWalkingDynamicField();
+            ContextCarrier contextCarrier = new ContextCarrier();
+            String topicName = requiredInfo.getTopic();
+            AbstractSpan activeSpan = ContextManager.createExitSpan(OPERATE_NAME_PREFIX + topicName +
+                    PRODUCER_OPERATE_NAME_SUFFIX, contextCarrier, requiredInfo.getServiceUrl());
+            Tags.MQ_BROKER.set(activeSpan, requiredInfo.getServiceUrl());
+            Tags.MQ_TOPIC.set(activeSpan, topicName);
+            SpanLayer.asMQ(activeSpan);
+            activeSpan.setComponent(ComponentsDefine.PULSAR_PRODUCER);
+            CarrierItem next = contextCarrier.items();
+            MessageImpl msg = (MessageImpl) allArguments[0];
+            while (next.hasNext()) {
+                next = next.next();
+                msg.getMessageBuilder().addProperties(PulsarApi.KeyValue.newBuilder()
+                        .setKey(next.getHeadKey())
+                        .setValue(next.getHeadValue()));
+            }
+            if (allArguments.length > 1) {
+                EnhancedInstance callbackInstance = (EnhancedInstance) allArguments[1];
+                if (callbackInstance != null) {
+                    ContextSnapshot snapshot = ContextManager.capture();
+                    if (null != snapshot) {
+                        SendCallbackEnhanceRequiredInfo callbackRequiredInfo = new SendCallbackEnhanceRequiredInfo();
+                        callbackRequiredInfo.setTopic(topicName);
+                        callbackRequiredInfo.setContextSnapshot(snapshot);
+                        callbackInstance.setSkyWalkingDynamicField(callbackRequiredInfo);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                              Object ret) throws Throwable {
+        if (allArguments.length > 0 && allArguments[0] != null) {
```

Comment:
I noticed there is plenty of `if (allArguments.length > 0 && allArguments[0] != null)`. Why do we need this? If we don't want to intercept any arg method, you could use the explicit method arg list in instrumentation definition.

Reply:
will update the instance method matcher